### PR TITLE
feat: Add actor audit log

### PR DESCRIPTION
**Pull request by FirestarJes**

### CODE REVIEW

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202310180956 Add ActorContact History.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    SET (SYSTEM_VERSIONING = ON (HISTORY_TABLE = dbo.[ActorContactHistory]));
+GO
+
+ALTER TABLE [dbo].[ActorContact]
+DROP CONSTRAINT DF_ChangedByIdentityId
+GO
+
+ALTER TABLE [dbo].[ActorContact]
+    ADD CONSTRAINT CHK_ActorContact_ChangedByIdentityId_NotEmpty CHECK (ChangedByIdentityId <> '00000000-0000-0000-0000-000000000000'),
+        CONSTRAINT CHK_ActorContact_DeletedByIdentityId_NotEmpty CHECK (DeletedByIdentityId <> '00000000-0000-0000-0000-000000000000');
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202310180956 Add ActorContact History.sql:25]: Does this work if DeletedBy.. is NULL?

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/ActorController.cs [hunk 2]
```diff
@@ -93,5 +93,18 @@ await _mediator
 
             return Ok();
         }
+
+        [HttpGet("{actorId:guid}/auditlogs")]
+        [AuthorizeUser(PermissionId.ActorsManage)]
+        public async Task<IActionResult> GetAuditLogsAsync(Guid actorId)
```
[the remark this answers is not in the document]
Comment (FirestarJes) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/ActorController.cs:99]: Oh, that is a mistake... will add.

[the remark this answers is not in the document]
Comment (FirestarJes) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/ActorController.cs:99]: Fixed

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202310180956 Add ActorContact History.sql (hunk 1 above)
Comment (FirestarJes) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202310180956 Add ActorContact History.sql:25]: image

Seems to work just fine with NULL

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/ActorContactAuditLogEntryRepository.cs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                        string.Empty,
+                        string.Empty));
+                }
+
+                // Check if the current entity is deleted, this can be done by verifying that we have no more entries for this category, and then add a deleted entry
+                if (current.Entity.DeletedByIdentityId != null && !HasNextForCategory(current.Entity, i, historicEntities))
+                {
+                    auditEntries.Add(new ActorContactAuditLogEntry(
+                        actorId,
+                        new AuditIdentity(current.Entity.ChangedByIdentityId),
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/ActorContactAuditLogEntryRepository.cs:89]: Should this be DeletedByIdentityId?

Comment (FirestarJes) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/ActorContactAuditLogEntryRepository.cs:89]: Indeed it should

Comment (FirestarJes) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/ActorContactAuditLogEntryRepository.cs:89]: Fixed

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/ActorContactAuditLogEntryRepository.cs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            for (var i = 0; i < historicEntities.Count; i++)
+            {
+                var isFirst = i == 0;
+                var current = historicEntities[i];
+                var previous = isFirst ? current : FindPreviousForCategory(current.Entity, i, historicEntities) ?? current;
+
+                // If this is the first time we see this category, we need to add the created entries
+                if (!categoriesSeen.Contains(current.Entity.Category))
+                {
+                    categoriesSeen.Add(current.Entity.Category);
```
[the remark this answers is not in the document]
Comment (FirestarJes) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/ActorContactAuditLogEntryRepository.cs:73]: HashSet doesn't have a TryAdd method, but i can change it to a dictionary instead?

[the remark this answers is not in the document]
Comment (FirestarJes) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/ActorContactAuditLogEntryRepository.cs]: Fixed

**Pull request merged**